presto-hive/src/main/java/io/prestosql/plugin/hive/orc/acid/DeletedRowsRegistry.java

```diff
@@ -0,0 +1,423 @@
+                missCount,
+                stats.loadSuccessCount(),
+                stats.loadExceptionCount(),
+                stats.totalLoadTime(),
+                stats.evictionCount());
+    }
+
+    public static long getCacheSize()
+    {
+        return cache.asMap().values().stream().map(rowIds -> rowIds.size()).reduce(0, Integer::sum) * ROW_ID_SIZE;
+    }
+
+    private class DeletedRowsLoader
+            implements Callable<List<RowId>>
+    {
+        private String partitionLocation;
+        private List<LocatedFileStatus> deleteDeltaFiles;
+        private String filename;
+        private Configuration configuration;
+        private DateTimeZone hiveStorageTimeZone;
+        private OrcPageSourceFactory pageSourceFactory;
+        private ConnectorSession session;
+
+        public DeletedRowsLoader(
+                String partitionLocation,
+                List<LocatedFileStatus> deleteDeltaFiles,
+                String filename,
+                OrcPageSourceFactory pageSourceFactory,
+                Configuration configuration,
+                ConnectorSession session,
+                DateTimeZone hiveStorageTimeZone)
+        {
+            this.partitionLocation = partitionLocation;
+            this.deleteDeltaFiles = deleteDeltaFiles;
+            this.filename = filename;
+            this.pageSourceFactory = pageSourceFactory;
+            this.configuration = configuration;
+            this.session = session;
+            this.hiveStorageTimeZone = hiveStorageTimeZone;
+        }
+
+        public String getPartitionLocation()
+        {
+            return partitionLocation;
+        }
+
+        public String getSplitFilename()
+        {
+            return filename;
+        }
+
+        @Override
+        public List<RowId> call()
+        {
+            ImmutableList.Builder<RowId> builder = ImmutableList.builder();
+            requestsNeedingLoad.incrementAndGet();
+            for (LocatedFileStatus fileStatus : deleteDeltaFiles) {
+                Optional<? extends ConnectorPageSource> pageSource = pageSourceFactory.createPageSource(
+                        configuration,
+                        session,
+                        fileStatus.getPath(),
+                        0,
+                        fileStatus.getLen(),
+                        fileStatus.getLen(),
+                        orcProperty,
+                        ACID_ROW_ID_COLUMN_HANDLES,
+                        TupleDomain.all(),
+                        hiveStorageTimeZone,
+                        Optional.empty());
+                if (!pageSource.isPresent()) {
+                    throw new PrestoException(
+                            HIVE_BAD_DATA,
+                            "Could not create page source for delete delta ORC file: " + fileStatus.getPath());
+                }
+                while (!pageSource.get().isFinished()) {
+                    Page page = pageSource.get().getNextPage();
```

Comment:
Hive ACID is designed so that loading all deleted rows into memory should not be necessary.
We may still choose to do so, but this deserves calling out in a comment explaining this design decision.

Reply:
cc @dain

Reply:
This doesnt load all deleted rows in memory all the time. It loads it in batches, deleted rows needed only for the running splits will be in cache and older ones will be evicted if needed.
Only when all splits are running in parallel, all deleted rows will be in memory.

Reply:
now,i can read orc table,but read repeat data when i update the table use hive,i should change where code

Reply:
@fengfengaaa i am not sure i understand. Also you're commenting under a closed PR.
May i suggest that we talk on our slack https://prestosql.io/slack.html ?